### PR TITLE
XGBoost: Create Booster parameters just once -> nicer log + single GPU check 

### PR DESCRIPTION
**Pull request by michalkurka**
This PR refactors the XGB implementation in order to create the parameters for the Booster just one time (at the beginning of the training).

This produces a little bit nicer logs (no repeated dumps of the parameters) and performs GPU check just one time.

### CODE REVIEW

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/BoosterParms.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+   * Iterates over a set of parameters and applies locale-specific formatting
+   * to decimal ones (Floats and Doubles).
+   *
+   * @param params Parameters to localize
+   * @return Map with localized parameter values
+   */
+  private static Map<String, Object> localizeDecimalParams(final Map<String, Object> params) {
+    Map<String, Object> localized = new HashMap<>(params.size());
+    final NumberFormat localizedNumberFormatter = DecimalFormat.getNumberInstance();
+    for (String key : params.keySet()) {
```
Comment (Pscheidl) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/BoosterParms.java]: Iterating over entrySet should be faster than first obtaining set of keys and then call get() for each key separately (even if the lookup it is constant). Lots of other methods calls.

**Comment by Pscheidl** [comment]
I created a PATCH that implements changes proposed to BoosterParms to make your work a little bit easier.
[boosterparms.txt](https://github.com/h2oai/h2o-3/files/1916976/boosterparms.txt)

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/BoosterParms.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  /**
+   * Iterates over a set of parameters and applies locale-specific formatting
+   * to decimal ones (Floats and Doubles).
+   *
+   * @param params Parameters to localize
+   * @return Map with localized parameter values
+   */
+  private static Map<String, Object> localizeDecimalParams(final Map<String, Object> params) {
+    Map<String, Object> localized = new HashMap<>(params.size());
+    final NumberFormat localizedNumberFormatter = DecimalFormat.getNumberInstance();
```
[the remark this answers is not in the document]
Comment (michalkurka) [reply h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/BoosterParms.java]: Instances of DecimalFormat don't guarantee thread-safety. Using them eg. as a static class member wouldn't be correct.

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/BoosterParms.java (hunk 1 above)
Comment (michalkurka) [reply h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/BoosterParms.java]: Please keep in mind that the backing map is an instance of NonBlockingHashMap.

Javadoc of NBHM for the entrySet() method says this:

    <p><strong>Warning:</strong> the iterator associated with this Set
    requires the creation of {@link java.util.Map.Entry} objects with each
    iteration.  The {@link NonBlockingHashMap} does not normally create or
    using {@link java.util.Map.Entry} objects so they will be created soley
    to support this iteration.  Iterating using {@link #keySet} or {@link
    #values} will be more efficient.

**Pull request merged**